### PR TITLE
Authorization for Downloads of restricted Bitstreams: Short lived token endpoint

### PR DESCRIPTION
**Pull request by KevinVdV**
## References
_Add references/links to any related tickets or PRs. These may include:_
* Link to [REST Contract](https://github.com/DSpace/Rest7Contract/pull/126)

## Description
This PR adds support for short lived token generation. The JWTTokenHandler class used to generate our session tokens has been split up into 2 parts:
* SessionJWTTokenHandler
* ShortLivedJWTTokenHandler

The difference being when that they have different configuration properties. If you have existing configuration that differs from the default ones you will need to look at the authentication.cfg file changes to see the new properties. Furthermore the "expiration" configuration property has been changed to use milliseconds instead of minutes.

## Instructions for Reviewers
If you want to test this PR locally alter the _jwt.shortLived.token.expiration_ in the local.cfg & put it to something like 30 seconds (as 2 seconds is very short if you want to test this).

To test this pr:
* Request a short lived token
* Provide the token to a private bitstream using the "token" parameter
* Wait until the configured time expires & try to download the file again, it should fail now

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by benbosman** [comment]
@KevinVdV I tried reviewing this PR, but the first thing I noticed is that my main authentication header is broken after creating a token.
This may indicate the eperson's session_salt is adjusted when the token is created, which should only happen for the main authentication header

Can you create an IT which:
* creates an authentication header using /server/api/authn/login
* creates a short-lived token using /server/api/authn/shortlivedtokens
* verifies the user is still authentication with the first authentication header using /server/api/authn/status

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:802]: can you add a test to verify that a valid short lived token provided for the user A that is not allowed to download the bitstream will be not sufficient for the download? i.e. I like to have a test to proof that is not sufficient to have a generic valid short lived token to download a restricted content

**Review by abollini — CHANGES_REQUESTED** [review]
hi @KevinVdV thanks for the PR it looks good overall.
I have made some small change request inline, the "bigger" one is related to the use of the word "session" to discriminate between short lived tokens and "normal" one. I suggest to name it just loginToken but it would be useful to get a feedback from @tdonohue on that to avoid unnecessary change if he disagree with my suggestion or multiple changes if he has a better name suggestion.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/jwt/JWTTokenRestAuthenticationServiceImpl.java
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/security/jwt/JWTTokenRestAuthenticationServiceImpl.java:54]: @KevinVdV : I agree with @abollini  here. The word "session" is not really appropriate here as it's confusing (is this a user session, an HTTP session, a Hibernate Session, or something else?).  I'd prefer we just call this `LoginJWTTokenHandler` as he suggested.

Similarly, all the new configurations should be `jwt.login.*` instead of `jwt.session.*` and similar.

**Comment by KevinVdV** [comment]
@abollini @benbosman All your feedback has been processed, let me know if anything is still missing (or if you find new things).

**Review by benbosman — CHANGES_REQUESTED** [review]
I have not reviewed this code yet, but I've extensively tested it, and this branch creates security risks.

Using a token it's possible to:
* Create another token (infinitely increasing the duration of the token by requesting a new one every second based on the previous token)
* Create another security header (creating a long-lived header based on a short-lived token)

I've tested this using:
```
token=$(curl --silent -X POST 'http://localhost:8080/server/api/authn/shortlivedtokens' -H "Authorization: $authorization2" | jq --raw-output '.token'); echo "$token"; curl -v "http://localhost:8080/server/api/authn/shortlivedtokens?token=$token" -X POST
token=$(curl --silent -X POST 'http://localhost:8080/server/api/authn/shortlivedtokens' -H "Authorization: $authorization2" | jq --raw-output '.token'); echo "$token"; curl -v "http://localhost:8080/server/api/authn/login?token=$token" -X POST
```

I would recommend to explicitly refuse /server/api/authn/login and /server/api/authn/shortlivedtokens to accept a token, and add ITs that verify it's not allowed (even with a valid token)

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:810]: this feedback seems to be not yet processed, can you double check please?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:840]: this feedback seems to be not yet processed, can you double check please?

**Review by abollini — CHANGES_REQUESTED** [review]
I have flagged as resolved my previous feedback were processed. I agree with the @benbosman finding but I'm also ok to postpone this security improvements to beta4. Indeed, in the Rest Contract discussion we have not identified the need to prevent the use of short lived token for the renew endpoint.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (jonas-atmire) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:840]: @abollini This is handled in the following test:
* https://github.com/atmire/DSpace/commit/27e733c1e8c557284bb756f32bc09c6a937bc4ee#diff-4906e304a90a1dcbf301b3896d39a353R848

This test goes through the following steps to verify the settings:
 * It sets the expiration of the token to 1ms on the configurationService
 * The code waits for 2ms
 * A login is tried with the previously retrieved token

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (jonas-atmire) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:810]: @abollini The following test verifies the short lived token validity:
* https://github.com/atmire/DSpace/commit/27e733c1e8c557284bb756f32bc09c6a937bc4ee#diff-4906e304a90a1dcbf301b3896d39a353R823

It goes through the following steps:
* A private bitstream is created
* It creates a user that does not have any particular permissions 
* The token for that user is retrieved and passed through to the bitstreams retrieval call
* The verification that the call is forbidden is done

**Comment by benbosman** [comment]
I've reviewed the code, and it seems all issues have already been solved, including the last 2 @abollini noted and @jonas-atmire responded to.
I also appreciate the large amount of tests.

Since I prefer not to create new security risks, I'd prefer to have that issue solved first. But if not realistic, we can perhaps postpone and create a critical ticket

**Comment by tdonohue** [comment]
@KevinVdV : would it be possible to resolve the security issues [noted by Ben](https://github.com/DSpace/DSpace/pull/2783#pullrequestreview-439235062) before merging this?  I'd rather not merge something with a known security issue. Just an FYI, we are looking to release beta3 as soon as this Thursday. So if this PR (and others dependent on it) is going to be in beta3, ideally we'd get this PR ready to merge by tomorrow (if that's at all possible).

**Comment by tdonohue** [comment]
@peter-atmire : It looks like your most recent commit here causes some unexpected behavior in the EPerson endpoints (based on the Travis CI results).  These tests are now throwing the wrong exception (a 403 instead of a 401) & therefore failing:
```[ERROR]   EPersonRestRepositoryIT.patchReplaceEmailWithTokenFail:2003 Status expected:<401> but was:<403>
[ERROR]   EPersonRestRepositoryIT.patchReplacePasswordWithOtherUserTokenFail:1964 Status expected:<401> but was:<403>
[ERROR]   EPersonRestRepositoryIT.patchReplacePasswordWithRandomTokenPatchFail:1916 Status expected:<401> but was:<403>
[ERROR]   EPersonRestRepositoryIT.registerNewAccountPatchUpdatePasswordRandomUserUuidFail:2053 Status expected:<401> but was:<403>
```
I suspect your change may be too low-level.  Maybe we should be rejecting the `token` param at the `AuthenticationRestController` level, rather than letting it through until it hits the `JWTTokenHandler`

**Comment by benbosman** [comment]
@tdonohue Because the Travis impact didn't match that commit, I've checked it out in more detail. The problem seems to be in the REST Contract, and not in the commit.
* Peter seems to have changed the AuthenticationRestControllerIT, which succeeds: https://travis-ci.com/github/DSpace/DSpace/builds/173680203#L5682
* The error https://travis-ci.com/github/DSpace/DSpace/builds/173680203#L5725 was quite odd, when I searched for that code in his branch, I couldn't find it. I think Travis merges this PR with master, and https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java#L1891 is failing
* The reason for this to fail can be found in the REST Contract. The https://github.com/DSpace/Rest7Contract/blob/master/epersons.md#create-new-eperson-based-on-registration-token uses a token parameter, and this PR also uses a token parameter. For the eperson, this implies the token is interpreted as an authentication token and as an eperson registration token. Because it's not a valid authentication token, the test are requested.
* This is a situation where the ITs have identified a conflict, we can't use the same token parameter for both the eperson registration and the general authentication. If Peter didn't commit this last change, Travis wouldn't have verified the merge of this PR with master, and the issue would have occurred in master. The problem will also occur with the code prior to Peter's last commit. I will create a REST Contract PR to adjust this

**Comment by tdonohue** [comment]
@benbosman : Thanks for looking into this.  I agree with your approach to simply rename one of these params. We don't want two different things named `token`.

**Review by benbosman — APPROVED** [review]
Thanks for the changes
I can see https://github.com/DSpace/DSpace/pull/2783/files/f9257dad12c97d1aa3daaf9beca35bed76ba98b0..9044daf50eb1b4044b6b4d80ce12b32e0a399712 solves the overlap of the tokens
I can also see https://github.com/DSpace/DSpace/pull/2783/commits/d364ac63837985f1b2162ac5e7e7cb6662cf3efd solves the security risk

I realize Travis is broken for all PRs today, but I've received the build output, and there are no issues

This has resolved all my feedback

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:811]: Verified this feedback looks to be resolved

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:843]: Verified this feedback has been addressed as well

**Review by tdonohue — APPROVED** [review]
👍 This looks good to me. I've reviewed the code, and don't have any additional questions/comments. I've also verified that all of @abollini 's prior feedback was addressed (he had asked for more tests which have been added). I've tested this in conjunction with the Angular PR https://github.com/DSpace/dspace-angular/pull/716 and they work together as described

**Comment by tdonohue** [comment]
Merging, as this is now at +2 and I've verified that all of @abollini 's prior feedback was addressed. Thanks @KevinVdV and @peter-atmire !

**Pull request merged**
